File: DataGridView ComboBox with ValueMember/[C#][VB]-DataGridView ComboBox with ValueMember/C# and VB/Example_vb/Extensions/Extensions.vb
''' <summary>
''' Several language extensions done so the code
''' in the form is cleaner as we could had done all
''' of this w/o extension methods.
''' </summary>
Module Extensions
    <System.Diagnostics.DebuggerStepThrough()>
    <System.Runtime.CompilerServices.Extension()>
    Public Function IsComboBoxCell(ByVal sender As DataGridViewCell) As Boolean
        Dim Result As Boolean = False
        If sender.EditType IsNot Nothing Then
            If sender.EditType Is GetType(DataGridViewComboBoxEditingControl) Then
                Result = True
            End If
        End If
        Return Result
    End Function
    <System.Diagnostics.DebuggerStepThrough()>
    <System.Runtime.CompilerServices.Extension()>
    Public Function ColorMessage(ByVal sender As BindingSource, ByVal ColorId As Integer) As String
        Dim index As Integer = sender.Find("Id", ColorId)
        If index = -1 Then
            ' this means someone messed with the database table
            Return ""
        Else
            Return CType(sender.Item(index), DataRowView).Row.Field(Of String)("Text")
        End If
    End Function
    <System.Diagnostics.DebuggerStepThrough()>
    <System.Runtime.CompilerServices.Extension()>
    Public Function ColorIdFromPerson(ByVal sender As BindingSource) As Integer
        Return CType(sender.Current, DataRowView).Row.Field(Of Integer)("ColorId")
    End Function
    <System.Diagnostics.DebuggerStepThrough()>
    <System.Runtime.CompilerServices.Extension()>
    Public Function ColorId(ByVal sender As BindingSource) As Integer
        Return CType(sender.Current, DataRowView).Row.Field(Of Integer)("ColorId")
    End Function
    <System.Diagnostics.DebuggerStepThrough()>
    <System.Runtime.CompilerServices.Extension()>
    Public Function Identifier(ByVal sender As BindingSource) As Integer
        Return CType(sender.Current, DataRowView).Row.Field(Of Integer)("Id")
    End Function
End Module

File: DataGridView ComboBox with ValueMember/[C#][VB]-DataGridView ComboBox with ValueMember/C# and VB/Example_cs/Exensions/Extensions.cs
using System.Data;
using System.Windows.Forms;

namespace Example_cs.Exensions
{
    static class Extensions
    {
        public static bool IsComboBoxCell(this DataGridViewCell sender)
        {
            bool Result = false;
            if (sender.EditType != null)
            {
                if (sender.EditType == typeof(DataGridViewComboBoxEditingControl))
                {
                    Result = true;
                }
            }
            return Result;
        }
        public static string ColorMessage(this BindingSource sender, int ColorId)
        {
            int index = sender.Find("Id", ColorId);
            if (index == -1)
            {
                // this means someone messed with the database table
                return "";
            }
            else
            {
                sender.Position = index;
                return ((DataRowView)sender.Current).Row.Field<string>("Text");
            }
        }
        public static int ColorIdFromPerson(this BindingSource sender)
        {
            return ((DataRowView)sender.Current).Row.Field<int>("ColorId");
        }
        public static int ColorId(this BindingSource sender)
        {
            return ((DataRowView)sender.Current).Row.Field<int>("ColorId");
        }
        public static int Identifier(this BindingSource sender)
        {
            return ((DataRowView)sender.Current).Row.Field<int>("Id");
        }
    }
}
